python/cudf/cudf/tests/test_dataframe.py

```diff
@@ -9063,6 +9063,152 @@ def test_dataframe_add_suffix():
+            },
+            ["id"],
+        ),
+        (
+            {
+                "id": [0, 0, 0, 0, 1, 1, 1],
+                "a": [10, 3, 4, 2, -3, 9, 10],
+                "b": [10, 23, -4, 2, -3, 9, 19],
+            },
+            ["id", "a"],
+        ),
+        (
+            {
+                "id": ["a", "a", "b", "b", "c", "c"],
+                "val": [np.nan, np.nan, np.nan, np.nan, np.nan, np.nan],
+            },
+            ["id"],
+        ),
+        ({"id": [1.0], "val1": [2.0], "val2": [3.0]}, ["id"]),
+    ],
+)
+@pytest.mark.parametrize(
+    "min_per", [1, 3, 4, 5],
+)
+@pytest.mark.parametrize(
+    "d_dof", [1, 3, 4, 5],
+)
+def test_groupby_covariance(data, gkey, min_per, d_dof):
+    gdf = cudf.DataFrame(data, nan_as_null=False)
+    pdf = gdf.to_pandas()
+
+    actual = gdf.groupby(gkey).cov(min_periods=min_per, ddof=d_dof)
+    expected = pdf.groupby(gkey).cov(min_periods=min_per, ddof=d_dof)
+
+    actual_no_args = gdf.groupby(gkey).cov()
+    expected_no_args = pdf.groupby(gkey).cov()
+
+    assert_eq(expected, actual)
+    assert_eq(expected_no_args, actual_no_args)
+
+
+def test_groupby_covariance_multiindex_dataframe():
+    gdf = cudf.DataFrame(
+        {
+            "a": [1, 1, 2, 2],
+            "b": [1, 1, 2, 3],
+            "c": [2, 3, 4, 5],
+            "d": [6, 8, 9, 1],
+        }
+    ).set_index(["a", "b"])
+
+    actual = gdf.groupby(level=["a", "b"]).cov(min_periods=0, ddof=1)
+    expected = (
+        gdf.to_pandas().groupby(level=["a", "b"]).cov(min_periods=0, ddof=1)
+    )
+
+    assert_eq(expected, actual)
+
+
+def test_groupby_covariance_empty_columns():
+    gdf = cudf.DataFrame(columns=["id", "val1", "val2"])
+    pdf = gdf.to_pandas()
+
+    actual = gdf.groupby("id").cov(min_periods=0, ddof=1)
+    expected = pdf.groupby("id").cov(min_periods=0, ddof=1)
+
+    assert_eq(
+        expected, actual, check_dtype=False, check_index_type=False,
+    )
+
+
+@pytest.mark.parametrize(
+    "data",
+    [
+        {
+            "id": ["a", "a", "a", "b", "b", "b", "c", "c", "c"],
+            "val1": ["v", "n", "k", "l", "m", "i", "y", "r", "w"],
+            "val2": ["d", "d", "d", "e", "e", "e", "f", "f", "f"],
+        },
+        {
+            "id": ["a", "a", "a", "b", "b", "b", "c", "c", "c"],
+            "val1": [1, 1, 1, 2, 2, 2, 3, 3, 3],
+            "val2": ["d", "d", "d", "e", "e", "e", "f", "f", "f"],
+        },
+    ],
+)
+@pytest.mark.parametrize("gkey", ["id", "val1", "val2"])
+def test_groupby_cov_invalid_column_types(data, gkey):
+    with pytest.raises(
+        TypeError, match="Covariance accepts only numerical column-pairs",
+    ):
+        cudf.DataFrame(data).groupby(gkey).cov(min_periods=0, ddof=1)
+
+
+def test_groupby_cov_positive_semidefinite_matrix():
+    # Refer to discussions in PR #9889 re "pair-wise deletion" strategy
+    # being used in pandas to compute the covariance of a dataframe with
+    # rows containing
```

Comment:
It looks like this sentence stopped early.
```suggestion
    # rows containing missing values.
```